iphone/CoreApi/CoreApi/Logger/Logger.mm

```diff
@@ -1,32 +1,192 @@
 #import "Logger.h"
-#import "base/logging.hpp"
+#import <OSLog/OSLog.h>
+
+#include "base/logging.hpp"
 
 @interface Logger ()
+
+@property (nullable, nonatomic) NSString * filePath;
+@property (nullable, nonatomic) NSFileHandle * fileHandle;
+@property (nonnull, nonatomic) os_log_t osLogger;
+@property (nonatomic) uint64_t fileSize;
+@property (class, readonly, nonatomic) dispatch_queue_t fileLoggingQueue;
+
++ (Logger *)logger;
++ (void)enableFileLogging;
++ (void)disableFileLogging;
++ (void)logMessageWithLevel:(base::LogLevel)level src:(base::SrcPoint const &)src message:(std::string const &)message;
 + (base::LogLevel)baseLevel:(LogLevel)level;
+
+- (BOOL)isFileLoggingEnabled;
+
 @end
 
+// Subsystem and category are used for the OSLog.
+NSString * const kLoggerSubsystem = [[NSBundle mainBundle] bundleIdentifier];
+NSString * const kLoggerCategory = @"OM";
+NSString * const kLogFileName = @"Log.log";
+NSUInteger const kMaxLogFileSize = 1024 * 1024 * 3; // 3 MB;
+
 @implementation Logger
 
-+ (void)log:(LogLevel)level message:(NSString*)message {
-  LOG_SHORT([Logger baseLevel:level], (message.UTF8String));
++ (void)initialize
+{
+  if (self == [Logger class]) {
+    SetLogMessageFn(&LogMessage);
+    SetAssertFunction(&AssertMessage);
+  }
+}
+
++ (Logger *)logger {
+  static Logger * logger = nil;
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
+    logger = [[self alloc] init];
+  });
+  return logger;
+}
+
++ (dispatch_queue_t)fileLoggingQueue {
+  static dispatch_queue_t fileLoggingQueue = nil;
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
+    dispatch_queue_attr_t attributes = dispatch_queue_attr_make_with_qos_class(DISPATCH_QUEUE_SERIAL, QOS_CLASS_UTILITY, 0);
+    fileLoggingQueue = dispatch_queue_create("app.organicmaps.fileLoggingQueue", attributes);
+  });
+  return fileLoggingQueue;
+}
+
+- (instancetype)init {
+  self = [super init];
+  if (self) {
+    _fileSize = 0;
+    _osLogger = os_log_create(kLoggerCategory.UTF8String, kLoggerSubsystem.UTF8String);
+  }
+  return self;
+}
+
+// MARK: - Public
+
++ (void)setFileLoggingEnabled:(BOOL)fileLoggingEnabled {
+  fileLoggingEnabled ? [self enableFileLogging] : [self disableFileLogging];
+}
+
++ (void)log:(LogLevel)level message:(NSString *)message {
+  LOG_SHORT([self baseLevel:level], (message.UTF8String));
 }
 
 + (BOOL)canLog:(LogLevel)level {
   return [Logger baseLevel:level] >= base::g_LogLevel;
 }
 
++ (nullable NSURL *)getLogFileURL {
+  Logger * logger = [self logger];
+  return [logger isFileLoggingEnabled] ? [NSURL fileURLWithPath:logger.filePath] : nil;
+}
+
++ (uint64_t)getLogFileSize {
+  return [self logger].fileSize;
+}
+
+// MARK: - C++ injection
+
+void LogMessage(base::LogLevel level, base::SrcPoint const & src, std::string const & message)
+{
+  [Logger logMessageWithLevel:level src:src message:message];
```

Comment:
1. Will it be simpler to call the default C++ LogMessage here first, and then call ObjC-specific log message?
2. Can a lambda (with necessary variables, e.g. ObjC logger instance) be set instead of a plain function?

Reply:
1. Whe the default c++ message func should be used here? If we call both we will duplicate logs in the debugger and it will produce an additional overhead.
2. Yes but it probably will be less readable. I use the same approach as in Android so the codebase will look more understandable for the other developers.

Reply:
1. Why logs will be duplicated? Now it works like this:
- This handler is called to log something
- This handler calls ObjC logMessageWithLevel
- logMessageWithLevel writes the default log (as in C++) and also writes to iOS 14+ only log and into a file

What I meant is:
- This handler is called to log something
- This handler calls default C++ log implementation
- This handler calls iOS-only ObjC logging implementation (that may also write to a file)

2. Won't using a lambda approach take fewer lines of code to implement/fewer function calls and be more clear/readable?

Reply:
> 1. Why logs will be duplicated? Now it works like this:
> 
> * This handler is called to log something
> * This handler calls ObjC logMessageWithLevel
> * logMessageWithLevel writes the default log (as in C++) and also writes to iOS 14+ only log and into a file

With this implementation we _do not use the default log_.
These cpp log helpers functions `WriteProlog` and `WriteLog` are used only for composing the output string and pass it into the os_log. They actually do not write anything.

The original function `LogMessageDefault` uses same approach to _compose_ message and then writes it:
https://github.com/organicmaps/organicmaps/blob/a2e9e38466daa62cc8060cd337b145faa432e538/base/logging.cpp#L67-L79

But we swap the `LogMessageDefault` with our obj implementation so no cpp logging is used anymore:
https://github.com/organicmaps/organicmaps/blob/a2e9e38466daa62cc8060cd337b145faa432e538/base/logging.cpp#L93-L97

Reply:
> What I meant is:
> 
> * This handler is called to log something
> * This handler calls default C++ log implementation
> * This handler calls iOS-only ObjC logging implementation (that may also write to a file)

I think that we should use the `os_log` instead of out custom cpp default log because it gives the ability to debug using the `Console app` without actual building and running the app - it may be very helpful to see the realtime logs in production build.

Reply:
> 2\. Won't using a lambda approach take fewer lines of code to implement/fewer function calls and be more clear/readable?

Are you sure that
```cpp
  if (self == [Logger class]) {
    base::SetLogMessageFn([] (base::LogLevel level, base::SrcPoint const & src, std::string const & msg) 
{
      [Logger logMessageWithLevel:level src:src message:msg];
      CHECK_LESS(level, base::g_LogAbortLevel, ("Abort. Log level is too serious", level));
    });
    base::SetAssertFunction([] (base::SrcPoint const & src, std::string const & msg) -> bool 
{
      [Logger logMessageWithLevel:LCRITICAL src:src message:msg];
      return true;
    });
  }
```

is clearer than 

```cpp
+ (void)initialize
{
  if (self == [Logger class]) {
    SetLogMessageFn(&LogMessage);
    SetAssertFunction(&AssertMessage);
  }
}

void LogMessage(base::LogLevel level, base::SrcPoint const & src, std::string const & message)
{
  [Logger logMessageWithLevel:level src:src message:message];
  CHECK_LESS(level, base::g_LogAbortLevel, ("Abort. Log level is too serious", level));
}

bool AssertMessage(base::SrcPoint const & src, std::string const & message)
{
  [Logger logMessageWithLevel:base::LCRITICAL src:src message:message];
  return true;
}
```
even if save 2 rows?